packages/common/src/hooks/useListData/useListData.ts

```diff
@@ -0,0 +1,75 @@
+import { SortBy } from './../useSortBy/useSortBy';
+import { UseMutateFunction } from 'react-query';
+import { useQueryClient, useMutation, useQuery } from 'react-query';
+import { QueryParams, useQueryParams } from '../useQueryParams';
+import { SortRule } from '../useSortBy';
+
+export interface ListApi<T> {
+  onQuery: ({
+    first,
+    offset,
+    sortBy,
+  }: {
+    first: number;
+    offset: number;
+    sortBy: SortBy<T>;
+  }) => () => Promise<{ data: T[]; totalLength: number }>;
+  onDelete: (toDelete: T[]) => Promise<void>;
+  onUpdate: (toUpdate: T) => Promise<T>;
+}
+
+interface ListDataState<T> extends QueryParams<T> {
+  data?: T[];
+  totalLength?: number;
+  fullQueryKey: readonly unknown[];
+  queryParams: QueryParams<T>;
+  onUpdate: UseMutateFunction<T, unknown, T, unknown>;
+  onDelete: UseMutateFunction<void, unknown, T[], unknown>;
+  isQueryLoading: boolean;
+  isUpdateLoading: boolean;
+  isDeleteLoading: boolean;
+  isLoading: boolean;
+  numberOfRows: number;
+}
+
+export const useListData = <T>(
+  initialSortBy: SortRule<T>,
+  queryKey: string | readonly unknown[],
+  api: ListApi<T>
+): ListDataState<T> => {
+  const queryClient = useQueryClient();
+  const { queryParams, first, offset, sortBy, numberOfRows } =
+    useQueryParams(initialSortBy);
+  const fullQueryKey = [queryKey, 'list', queryParams];
+
+  const { data, isLoading: isQueryLoading } = useQuery(
+    fullQueryKey,
+    api.onQuery({ first, offset, sortBy })
+  );
+
+  const invalidation = () => queryClient.invalidateQueries(queryKey);
+
+  const { mutateAsync: onDelete, isLoading: isDeleteLoading } = useMutation(
+    api.onDelete,
+    { onSuccess: invalidation }
+  );
+
+  const { mutateAsync: onUpdate, isLoading: isUpdateLoading } = useMutation(
+    api.onUpdate,
+    { onSuccess: invalidation }
+  );
+
+  return {
```

Comment:
is super simple right now but by just defining a few graphql queries in an `api` object, gives a lot of functionality.

- `useMutation` -> `useOptimisticMutation`, probably
- error handling
- Likely need to return some more fields from the mutations
- Trying to think of a way where we can also call `useListData` with the same QueryKey, in any random component and get all the same data like you already can with `useQuery`, but this might need a separate hook 🤔 

Also trying to be conscious that putting too much here is going to make it very confusing and hard to maintain.. that is why, I suppose, I have tried to break it down a lot..